File: src/queue/queue-engine.ts
import * as fs from 'fs';
import * as crypto from 'crypto';
import { queuePath } from '../utils/paths';
import type { QueueEntry, QueueFile } from '../types';

/**
 * Atomically write a file.
 */
function atomicWrite(filePath: string, content: string): void {
  const tmp = filePath + '.tmp.' + process.pid;
  fs.writeFileSync(tmp, content, 'utf-8');
  fs.renameSync(tmp, filePath);
}

/**
 * Initialize an empty queue for a terminal.
 */
export function initQueue(terminalId: string, projectRoot: string): void {
  const filePath = queuePath(terminalId, projectRoot);
  const queue: QueueFile = { terminal_id: terminalId, entries: [] };
  atomicWrite(filePath, JSON.stringify(queue, null, 2));
}

/**
 * Read the queue for a terminal.
 */
export function readQueue(terminalId: string, projectRoot: string): QueueFile {
  const filePath = queuePath(terminalId, projectRoot);

  if (!fs.existsSync(filePath)) {
    return { terminal_id: terminalId, entries: [] };
  }

  try {
    const raw = fs.readFileSync(filePath, 'utf-8');
    return JSON.parse(raw) as QueueFile;
  } catch {
    return { terminal_id: terminalId, entries: [] };
  }
}

/**
 * Write the queue atomically.
 */
export function writeQueue(terminalId: string, queue: QueueFile, projectRoot: string): void {
  const filePath = queuePath(terminalId, projectRoot);
  atomicWrite(filePath, JSON.stringify(queue, null, 2));
}

/**
 * Enqueue a normal input.
 */
export function enqueueInput(
  terminalId: string,
  input: string,
  priority: number,
  projectRoot: string
): QueueEntry {
  const queue = readQueue(terminalId, projectRoot);

  const entry: QueueEntry = {
    queue_id: crypto.randomUUID(),
    input,
    priority,
    mode: 'normal',
    status: 'queued',
    created_at: new Date().toISOString(),
    sent_at: null,
  };

  queue.entries.push(entry);
  writeQueue(terminalId, queue, projectRoot);

  return entry;
}

/**
 * Enqueue an override input — cancels all pending inputs first.
 */
export function enqueueOverride(
  terminalId: string,
  input: string,
  projectRoot: string
): { entry: QueueEntry; cancelled: number } {
  const queue = readQueue(terminalId, projectRoot);

  // Cancel all pending (unsent) entries
  let cancelled = 0;
  for (const entry of queue.entries) {
    if (entry.status === 'queued') {
      entry.status = 'cancelled';
      cancelled++;
    }
  }

  // Insert override at the beginning
  const entry: QueueEntry = {
    queue_id: crypto.randomUUID(),
    input,
    priority: Number.MAX_SAFE_INTEGER,
    mode: 'override',
    status: 'queued',
    created_at: new Date().toISOString(),
    sent_at: null,
  };

  queue.entries.push(entry);
  writeQueue(terminalId, queue, projectRoot);

  return { entry, cancelled };
}

/**
 * Get the next queued entry sorted by priority DESC, created_at ASC.
 */
export function getNextQueued(terminalId: string, projectRoot: string): QueueEntry | null {
  const queue = readQueue(terminalId, projectRoot);

  const pending = queue.entries.filter((e) => e.status === 'queued');
  if (pending.length === 0) return null;

  // Sort: priority DESC, then created_at ASC
  pending.sort((a, b) => {
    if (b.priority !== a.priority) return b.priority - a.priority;
    return new Date(a.created_at).getTime() - new Date(b.created_at).getTime();
  });

  return pending[0];
}

/**
 * Mark a queue entry as sent.
 */
export function markSent(terminalId: string, queueId: string, projectRoot: string): void {
  const queue = readQueue(terminalId, projectRoot);

  const entry = queue.entries.find((e) => e.queue_id === queueId);
  if (entry) {
    entry.status = 'sent';
    entry.sent_at = new Date().toISOString();
    writeQueue(terminalId, queue, projectRoot);
  }
}

/**
 * Get pending queue length.
 */
export function pendingCount(terminalId: string, projectRoot: string): number {
  const queue = readQueue(terminalId, projectRoot);
  return queue.entries.filter((e) => e.status === 'queued').length;
}

File: python/clrun/queue/queue_engine.py
"""Priority-based input queue with FIFO ordering and override mode."""

from __future__ import annotations

import json
import os
import uuid
from datetime import datetime, timezone

from clrun.types import QueueEntry, QueueFile
from clrun.utils.paths import queue_path


def _atomic_write(filepath: str, content: str) -> None:
    tmp = filepath + f".tmp.{os.getpid()}"
    with open(tmp, "w", encoding="utf-8") as f:
        f.write(content)
    os.replace(tmp, filepath)


def init_queue(terminal_id: str, project_root: str) -> None:
    fp = queue_path(terminal_id, project_root)
    q = QueueFile(terminal_id=terminal_id)
    _atomic_write(fp, json.dumps(q.to_dict(), indent=2))


def read_queue(terminal_id: str, project_root: str) -> QueueFile:
    fp = queue_path(terminal_id, project_root)
    if not os.path.exists(fp):
        return QueueFile(terminal_id=terminal_id)
    try:
        with open(fp, "r", encoding="utf-8") as f:
            return QueueFile.from_dict(json.load(f))
    except Exception:
        return QueueFile(terminal_id=terminal_id)


def write_queue(terminal_id: str, queue: QueueFile, project_root: str) -> None:
    fp = queue_path(terminal_id, project_root)
    _atomic_write(fp, json.dumps(queue.to_dict(), indent=2))


def enqueue_input(terminal_id: str, text: str, priority: int, project_root: str) -> QueueEntry:
    queue = read_queue(terminal_id, project_root)
    entry = QueueEntry(
        queue_id=str(uuid.uuid4()),
        input=text,
        priority=priority,
        mode="normal",
        status="queued",
        created_at=datetime.now(timezone.utc).isoformat(),
    )
    queue.entries.append(entry)
    write_queue(terminal_id, queue, project_root)
    return entry


def enqueue_override(terminal_id: str, text: str, project_root: str) -> tuple[QueueEntry, int]:
    queue = read_queue(terminal_id, project_root)
    cancelled = 0
    for e in queue.entries:
        if e.status == "queued":
            e.status = "cancelled"
            cancelled += 1
    entry = QueueEntry(
        queue_id=str(uuid.uuid4()),
        input=text,
        priority=2**53,
        mode="override",
        status="queued",
        created_at=datetime.now(timezone.utc).isoformat(),
    )
    queue.entries.append(entry)
    write_queue(terminal_id, queue, project_root)
    return entry, cancelled


def get_next_queued(terminal_id: str, project_root: str) -> QueueEntry | None:
    queue = read_queue(terminal_id, project_root)
    pending = [e for e in queue.entries if e.status == "queued"]
    if not pending:
        return None
    pending.sort(key=lambda e: (-e.priority, e.created_at))
    return pending[0]


def mark_sent(terminal_id: str, queue_id: str, project_root: str) -> None:
    queue = read_queue(terminal_id, project_root)
    for e in queue.entries:
        if e.queue_id == queue_id:
            e.status = "sent"
            e.sent_at = datetime.now(timezone.utc).isoformat()
            break
    write_queue(terminal_id, queue, project_root)


def pending_count(terminal_id: str, project_root: str) -> int:
    queue = read_queue(terminal_id, project_root)
    return sum(1 for e in queue.entries if e.status == "queued")
